security/msal/MsalClient.ts

```diff
@@ -247,6 +267,82 @@ export class MsalClient extends BaseOAuthClient<MsalClientConfig, MsalTokenSpec>
             : await client.logoutPopup(opts);
     }
 
+    //------------------------
+    // Telemetry
+    //------------------------
+    enableTelemetry(): void {
+        if (this._telemetryCbHandle) {
+            this.logInfo('Telemetry already enabled', this.telemetryResults);
+            return;
+        }
+
+        this.telemetryResults = {
+            startTime: new Date(),
+            endTime: null,
+            events: {}
+        };
+
+        this._telemetryCbHandle = this.client.addPerformanceCallback(events => {
+            events.forEach(e => {
+                try {
+                    const {events} = this.telemetryResults,
+                        {name, startTimeMs, durationMs, success, errorName, errorCode} = e,
+                        eTime = startTimeMs ? new Date(startTimeMs) : new Date();
+
+                    if (!events[name]) {
+                        events[name] = {
+                            firstTime: eTime,
+                            lastTime: eTime,
+                            successCount: 0,
+                            failureCount: 0,
+                            duration: {count: 0, total: 0, average: 0, worst: 0},
+                            lastFailure: null
+                        };
+                    }
+
+                    const eResult = events[name];
+                    eResult.lastTime = eTime;
+
+                    if (success) {
+                        eResult.successCount++;
+                    } else {
+                        eResult.failureCount++;
+                        eResult.lastFailure = {
+                            time: eTime,
+                            code: errorCode,
+                            name: errorName,
+                            raw: e
+                        };
+                    }
+
+                    const {duration} = eResult;
+                    if (durationMs) {
```

Comment:
is zero milliseconds ever a possibility for a no-op cache lookup?   safer would be isFinite()

Reply:
I imagined it could be, but was also deliberately skipping those cases on the theory that it might be useful to track events-with-non-zero-duration only in this nested block